File: day-frontend/src/types/cleaning.ts
export type CleaningType = 'post_checkout' | 'mid_stay' | 'on_demand'
export type CleaningStatus = 'pending' | 'assigned' | 'in_progress' | 'done' | 'verified'
export type ReportStatus = 'submitted' | 'approved' | 'rejected'
export type RoomType = 'bathroom' | 'kitchen' | 'bedroom' | 'other'

export interface CleaningTask {
  id: string
  company_id: string
  property_id: string
  booking_id: string | null
  cleaner_id: string | null
  type: CleaningType
  status: CleaningStatus
  scheduled_date: string | null
  scheduled_time: string | null
  notes: string | null
  started_at: string | null
  completed_at: string | null
  verified_at: string | null
  property_name: string | null
  property_internal_name: string | null
  created_at: string
  updated_at: string
}

export interface CleaningTaskCreateInput {
  property_id: string
  type?: CleaningType
  booking_id?: string
  cleaner_id?: string
  scheduled_date?: string
  scheduled_time?: string
  notes?: string
}

export interface CleaningTaskListResponse {
  items: CleaningTask[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface CleaningTaskFilters {
  page?: number
  per_page?: number
  status?: CleaningStatus
  property_id?: string
  cleaner_id?: string
  date_from?: string
  date_to?: string
}

// Checklist Template
export interface ChecklistTemplate {
  id: string
  company_id: string
  name: string
  created_at: string
}

export interface ChecklistItem {
  id: string
  template_id: string
  title: string
  sort_order: number
}

export interface ChecklistTemplateDetail {
  template: ChecklistTemplate
  items: ChecklistItem[]
}

export interface ChecklistTemplateCreateInput {
  name: string
  items?: { title: string; sort_order?: number }[]
}

export interface ChecklistTemplateUpdateInput {
  name: string
}

export interface ChecklistItemUpdateInput {
  title: string
}

// Cleaning Report
export interface CleaningReport {
  id: string
  task_id: string
  cleaner_id: string
  status: ReportStatus
  notes: string | null
  submitted_at: string | null
  created_at: string
}

export interface ReportPhoto {
  id: string
  report_id: string
  url: string
  room_type: RoomType
  metadata: Record<string, unknown> | null
  metadata_verified: boolean
}

export interface ReportChecklist {
  id: string
  report_id: string
  checklist_item_id: string
  is_done: boolean
  note: string | null
}

export interface CleaningReportDetail {
  report: CleaningReport
  photos: ReportPhoto[]
  checklist: ReportChecklist[]
}

export interface CleaningTaskDetail {
  task: CleaningTask
  report: CleaningReportDetail | null
}

export interface SubmitReportInput {
  cleaner_id: string
  notes?: string
  photos?: { url: string; room_type?: RoomType; metadata?: Record<string, unknown> }[]
  checklist?: { checklist_item_id: string; is_done: boolean; note?: string }[]
}

// Cleaner Rating
export interface CleanerRating {
  id: string
  company_id: string
  cleaner_id: string
  task_id: string | null
  rated_by: string | null
  score: number
  review: string | null
  kpi_metrics: Record<string, unknown> | null
  created_at: string
}

export interface RateCleanerInput {
  cleaner_id: string
  score: number
  task_id?: string
  review?: string
}

export interface CleanerKPI {
  cleaner_id: string
  avg_score: number
  total_ratings: number
  recent_ratings: {
    id: string
    score: number
    review: string | null
    created_at: string | null
  }[]
}

// State machine transitions
export const CLEANING_VALID_TRANSITIONS: Record<CleaningStatus, CleaningStatus[]> = {
  pending: ['assigned'],
  assigned: ['in_progress'],
  in_progress: ['done'],
  done: ['verified'],
  verified: [],
}

File: day-backend/app/presentation/schemas/cleaning.py
from __future__ import annotations

import uuid
from datetime import date, datetime, time

from pydantic import BaseModel, ConfigDict, Field

from app.domain.cleaning.value_objects import (
    CleaningStatus,
    CleaningType,
    ReportStatus,
    RoomType,
)

# ---------- Cleaning Task ----------


class CleaningTaskCreate(BaseModel):
    property_id: uuid.UUID
    type: CleaningType = CleaningType.POST_CHECKOUT
    booking_id: uuid.UUID | None = None
    cleaner_id: uuid.UUID | None = None
    scheduled_date: date | None = None
    scheduled_time: time | None = None
    notes: str | None = None


class CleaningTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    property_id: uuid.UUID
    booking_id: uuid.UUID | None = None
    cleaner_id: uuid.UUID | None = None
    type: CleaningType
    status: CleaningStatus
    scheduled_date: date | None = None
    scheduled_time: time | None = None
    notes: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    verified_at: datetime | None = None
    property_name: str | None = None
    property_internal_name: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class CleaningTaskStatusChange(BaseModel):
    target_status: CleaningStatus


class CleaningTaskAssign(BaseModel):
    cleaner_id: uuid.UUID


class CleaningTaskListResponse(BaseModel):
    items: list[CleaningTaskResponse]
    total: int
    page: int
    per_page: int
    pages: int


# ---------- Checklist Template ----------


class ChecklistItemInput(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    sort_order: int = 0


class ChecklistTemplateCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    items: list[ChecklistItemInput] = []


class ChecklistTemplateUpdate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)


class ChecklistItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    template_id: uuid.UUID
    title: str
    sort_order: int


class ChecklistTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    name: str
    created_at: datetime | None = None


class ChecklistTemplateDetailResponse(BaseModel):
    template: ChecklistTemplateResponse
    items: list[ChecklistItemResponse]


class ChecklistItemAdd(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)
    sort_order: int | None = None


class ChecklistItemUpdate(BaseModel):
    title: str = Field(..., min_length=1, max_length=500)


class ChecklistItemReorder(BaseModel):
    item_ids: list[uuid.UUID] = Field(..., min_length=1)


# ---------- Cleaning Report ----------


class ReportPhotoInput(BaseModel):
    url: str = Field(..., min_length=1, max_length=1024)
    room_type: RoomType = RoomType.OTHER
    metadata: dict | None = None


class ReportChecklistInput(BaseModel):
    checklist_item_id: uuid.UUID
    is_done: bool = False
    note: str | None = None


class SubmitReportCreate(BaseModel):
    cleaner_id: uuid.UUID
    notes: str | None = None
    photos: list[ReportPhotoInput] = []
    checklist: list[ReportChecklistInput] = []


class ReportPhotoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    report_id: uuid.UUID
    url: str
    room_type: RoomType
    metadata: dict | None = None
    metadata_verified: bool


class ReportChecklistResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    report_id: uuid.UUID
    checklist_item_id: uuid.UUID
    is_done: bool
    note: str | None = None


class CleaningReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    task_id: uuid.UUID
    cleaner_id: uuid.UUID
    status: ReportStatus
    notes: str | None = None
    submitted_at: datetime | None = None
    created_at: datetime | None = None


class CleaningReportDetailResponse(BaseModel):
    report: CleaningReportResponse
    photos: list[ReportPhotoResponse]
    checklist: list[ReportChecklistResponse]


# ---------- Task Detail ----------


class CleaningTaskDetailResponse(BaseModel):
    task: CleaningTaskResponse
    report: CleaningReportDetailResponse | None = None


# ---------- Cleaner Rating ----------


class RateCleanerCreate(BaseModel):
    cleaner_id: uuid.UUID
    score: int = Field(..., ge=1, le=5)
    task_id: uuid.UUID | None = None
    review: str | None = None


class CleanerRatingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_id: uuid.UUID
    cleaner_id: uuid.UUID
    task_id: uuid.UUID | None = None
    rated_by: uuid.UUID | None = None
    score: int
    review: str | None = None
    kpi_metrics: dict | None = None
    created_at: datetime | None = None


class CleanerKPIResponse(BaseModel):
    cleaner_id: str
    avg_score: float
    total_ratings: int
    recent_ratings: list[dict]
